addons/hw_drivers/server_logger.py

```diff
@@ -0,0 +1,164 @@
+        while self._odoo_server_url and self._active:  # allow to exit the loop on thread.join
+            time.sleep(self._FLUSH_INTERVAL)
+            self._flush_logs(odoo_session)
+
+    def _flush_logs(self, odoo_session):
+        def convert_to_byte(s):
+            return bytes(s, encoding="ascii") + b'<log/>\n'
+
+        def convert_server_line(log_level, line_formatted):
+            return convert_to_byte(f"{log_level},{line_formatted}")
+
+        def empty_queue():
+            yield convert_to_byte(f"mac {helpers.get_mac_address()}")
+            for _ in range(self._MAX_BATCH_SIZE):
+                # Use a limit to avoid having too heavy requests & infinite loop of the queue receiving new entries
+                try:
+                    log_record = self._log_queue.get_nowait()
+                    yield convert_server_line(log_record.levelno, self.format(log_record))
+                except queue.Empty:
+                    break
+
+            # Report to the server if the queue is close from saturation
+            if queue_size >= .8 * self._MAX_QUEUE_SIZE:
+                log_message = "The IoT {} queue is saturating: {}/{} ({:.2f}%)".format(
+                    self.__class__.__name__, queue_size, self._MAX_QUEUE_SIZE,
+                    100 * queue_size / self._MAX_QUEUE_SIZE)
+                _logger.warning(log_message)  # As we don't log our own logs, this will be part of the IoT logs
+                # In order to report this to the server (on the current batch) we will append it manually
+                yield convert_server_line(logging.WARNING, log_message)
+
+        queue_size = self._log_queue.qsize()  # This is an approximate value
+
+        if not self._odoo_server_url or queue_size == 0:
+            return
+        try:
+            odoo_session.post(
+                self._odoo_server_url + '/iot/log',
+                data=empty_queue(),
+                timeout=self._REQUEST_TIMEOUT
+            ).raise_for_status()
+            self._next_disconnection_time = None
+        except (requests.exceptions.ConnectTimeout, requests.exceptions.ConnectionError, urllib3.exceptions.NewConnectionError) as request_errors:
+            now = time.time()
+            if not self._next_disconnection_time or now >= self._next_disconnection_time:
+                _logger.info("Connection with the server to send the logs failed. It is likely down: %s", request_errors)
+                self._next_disconnection_time = now + self._DELAY_BEFORE_NO_SERVER_LOG
+        except Exception as _:
+            _logger.exception('Unexpected error happened while sending logs to server')
+
+    def emit(self, record):
+        # This is important that this method is as fast as possible.
+        # The log calls will be waiting for this function to finish
+        if not self._active:
+            return
+        try:
+            self._log_queue.put_nowait(record)
+        except queue.Full:
+            pass
```

Comment:
tell me if I understand correctly: you try/except the full queue instead of using Queue.full() method to make it faster?

Reply:
not really a question of speed, `put_nowait` can raise this exception if it is Full, so for reading  convenience it does make sense rather than a if not queue.full()
furthermore the documentation state, https://docs.python.org/3/library/queue.html#queue.Queue.full:
`if full() returns False it doesn’t guarantee that a subsequent call to put() will not block.`
even if here it is not blocking anyway.
I guess the final result would be the same, but I prefer this approach as it make more sense in my opinion